src/main/scala/netty/ActorChannelConnector.scala

```diff
@@ -5,34 +5,81 @@ import io.netty.buffer.Unpooled
-import org.apache.pekko.actor.typed.ActorRef
+import org.apache.pekko.actor.typed.{ ActorRef, Scheduler }
 
 import lila.ws.Controller.Endpoint
 import lila.ws.netty.ProtocolHandler.key
 
-final private class ActorChannelConnector(clients: ActorRef[Clients.Control])(using Executor):
+final private class ActorChannelConnector(
+    clients: ActorRef[Clients.Control],
+    config: com.typesafe.config.Config,
+    settings: util.SettingStore
+)(using
+    scheduler: Scheduler,
+    ec: Executor
+):
+  private val step     = intSetting("netty.flush.step")
+  private val interval = intSetting("netty.flush.interval-millis")
+  private val maxDelay = intSetting("netty.flush.max-delay-millis")
+  private val flushQ   = new java.util.concurrent.ConcurrentLinkedQueue[Channel]()
+  private val monitor  = Monitor.connector.flush
+
+  scheduler.scheduleOnce(1 second, () => flush())
+  scheduler.scheduleWithFixedDelay(1 minute, 1 minute): () =>
+    monitor.config.step.update(step.get())
+    monitor.config.interval.update(interval.get())
+    monitor.config.maxDelay.update(maxDelay.get())
 
   def apply(endpoint: Endpoint, channel: Channel): Unit =
     val clientPromise = Promise[Client]()
     channel.attr(key.client).set(clientPromise.future)
-    val channelEmit = emitToChannel(channel)
-    val monitoredEmit: ClientEmit = (msg: ipc.ClientIn) =>
-      endpoint.emitCounter.increment()
-      channelEmit(msg)
-    clients ! Clients.Control.Start(endpoint.behavior(monitoredEmit), clientPromise)
+    val channelEmit: ClientEmit =
+      val emitter = emitToChannel(channel, withFlush = endpoint.alwaysFlush)
+      (msg: ipc.ClientIn) =>
+        endpoint.emitCounter.increment()
+        emitter(msg)
+    clients ! Clients.Control.Start(endpoint.behavior(channelEmit), clientPromise)
     channel.closeFuture.addListener:
       new GenericFutureListener[NettyFuture[Void]]:
         def operationComplete(f: NettyFuture[Void]): Unit =
           channel.attr(key.client).get.foreach { client =>
             clients ! Clients.Control.Stop(client)
           }
 
-  private def emitToChannel(channel: Channel): ClientEmit =
+  private def intSetting(key: String) =
+    settings.makeSetting(key, config.getInt(key))
+
+  private def emitToChannel(channel: Channel, withFlush: Boolean): ClientEmit =
     case ipc.ClientIn.Disconnect(reason) =>
       channel
         .writeAndFlush(CloseWebSocketFrame(WebSocketCloseStatus(4010, reason)))
         .addListener(ChannelFutureListener.CLOSE)
     case ipc.ClientIn.RoundPingFrameNoFlush =>
       channel.write { PingWebSocketFrame(Unpooled.copyLong(System.currentTimeMillis())) }
-    case in =>
+    case in if withFlush || interval.get() <= 0 =>
       channel.writeAndFlush(TextWebSocketFrame(in.write))
+    case in =>
+      channel.write(TextWebSocketFrame(in.write))
+      flushQ.add(channel)
+
+  private def flush(): Unit =
+    val qSize           = flushQ.size
```

Comment:
Feels like if interval is expected to be 0 sometimes, you shouldn't be calling <queue>.size in those cases, as it's a linear op with lots of memory barriers

Reply:
when `interval.get() == 0` then `flushQ.add(channel)` is not called and the queue remains empty. It's not obvious tho, feels like an abstraction is missing to make the flow clearer.